### ISSUE TITLE
Support for \Model\changed() - ?

### ISSUE THREAD

**Issue by module0x90** [ikkez/f3-cortex#36]
Hi,

when migrating from F3 \DB\SQL\Mapper to \DB\Cortex I realized that changed() has disappeared.
I understand the Collection class has hasChanged() but with my limited PHP knowledge this doesn't appear to be working for me.
Basically I am doing the classical 
1) Display form with POST. vars 2) Hydrate with copyFrom(POST) 3) changed()?

How can this be used with Cortex?

Cheers

Thomas

**Comment by jekabsmilbrets** [comment]
Have you tried - $Model->mapper->changed('name');

?

**Comment by module0x90** [comment]
Ah .... gotcha.

Thank you.
--
www​.​preissler​.​co​.​uk | Twitter: @module0x90
GPG: BA359D78200264B363314AF5E3839138A11FFD2A

February 7 2017 9:58 AM, "Jēkabs Milbrets"  wrote:
	Have you tried - $Model->mapper->changed('name');

	?

**Comment by module0x90** [comment]
Now I am getting 

Call to a member function changed() on null

And my code is

 if (! $this->resource->mapper->changed()) {

whereas

$this->resource = new ModelMapper

Can this actually work? "mapper" is declared as "protected" inside the Class?
Cheers

Thomas

--
www​.​preissler​.​co​.​uk | Twitter: @module0x90
GPG: BA359D78200264B363314AF5E3839138A11FFD2A

February 7 2017 12:05 PM, "Thomas Preissler"  wrote:

Ah .... gotcha.

Thank you.
--
www​.​preissler​.​co​.​uk | Twitter: @module0x90
GPG: BA359D78200264B363314AF5E3839138A11FFD2A

February 7 2017 9:58 AM, "Jēkabs Milbrets"  wrote:
	Have you tried - $Model->mapper->changed('name');

	?

**Comment by jekabsmilbrets** [comment]
My bad. I'm sorry, I just realised that I used this method with-in base Model that extends  Cortex ( in a beforeupdate hook).

**Comment by ikkez** [comment]
There is no `changed` method yet, because currently it's only the SQL mapper that has that changed-method built-in. the Mongo and Jig mapper still lacks of that feature. I could re-implement that feature directly into cortex, or initiate that it'll be added to the F3 core and then add a pass-through to those in cortex. what do you think?

**Comment by module0x90** [comment]
Considering Cortex already has copyfrom() (as well as db/sql/mapper.php) it would be an idea to have this in Cortex. I see the changed() feature what can be completely independent of the underlaying DB. Comparing copyfrom() in db/sql/mapper.php with cortex.php it's quite "similar"...
But this would obviously leave non-Cortex users out in the rain... so alternatively implementing it in F3 and then just do a pass-through in Cortex would suit everyone.

For my case I had to implement some Audit() feature, ie. I record DB record creation and the delta of changes in a separate table. I realized I can easily implement my own changed() by comparing originalData (what's filled on onload()) with the current data via cast(), done it and working already :-)
Just don't do a cast() inside onload()... this eats your memory.
